File: functions/python/computeEoForRho.py
import numpy as np

def computeEoForRho(rho, Q, pi_matrix, g_matrix):
    """
    Computes E0(ρ) for a given ρ value, used in the optimization of the error exponent 
    E(R) = max{E0(ρ) - ρR} for communication systems.
    
    Parameters:
        rho (float): Value of ρ parameter, 0 ≤ ρ ≤ 1.
        Q (np.ndarray): 1D array of probabilities for constellation points (length M).
        pi_matrix (np.ndarray): Matrix of quadrature weight products with shape (M, K),
                                where K is typically related to the number of nodes.
        g_matrix (np.ndarray): Matrix of G function values with shape (M, K).
    
    Returns:
        float: Computed value of E0(ρ) for the given parameters.
    """
    # Compute the intermediate terms.
    # Each element of g_matrix is raised to the power of 1/(1+rho)
    g_powered = g_matrix ** (1 / (1 + rho))
    
    # Weighted sum over g_powered. Here Q is a 1D array of length M and g_powered is (M, K)
    # so the dot product gives a 1D array of length K.
    qg_rho = np.dot(Q, g_powered)
    
    # Raise the weighted sum to the power of rho
    qg_rho_t = qg_rho ** rho
    
    # Compute the Pi-weighted terms.
    pi_g_rho = pi_matrix * (g_matrix ** (-rho / (1 + rho)))
    
    # Combine results:
    # First, perform matrix multiplication between pi_g_rho (M, K) and qg_rho_t (K,)
    # which results in a vector of length M; then take the dot product with Q (length M)
    # to obtain a scalar.
    component_second = np.dot(Q, np.dot(pi_g_rho, qg_rho_t))
    
    # Compute the final value for E0(rho) using base-2 logarithm.
    Eo_rho = -np.log2((1 / np.pi) * component_second)
    
    return Eo_rho

File: functions/computeEoForRho.m
function Eo_rho = computeEoForRho(rho, Q, pi_matrix, g_matrix)
% computeEoForRho - Computes E0(ρ) for a given ρ value in the optimization of 
% error exponent E(R) = max{E0(ρ) - ρR} for communication systems
%
% Inputs:
%   rho       - Value of ρ parameter, 0 ≤ ρ ≤ 1
%   Q         - Probability distribution of constellation points (1×M)
%   pi_matrix - Matrix of quadrature weights products (M×N²M)
%   g_matrix  - Matrix of G function values for each combination (M×N²M)
%
% Output:
%   Eo_rho    - Computed value of E0(ρ) for the given parameters

    % Compute the intermediate terms
    g_powered = g_matrix.^(1 / (1 + rho));
    qg_rho = Q' * g_powered;     % Weighted sum over g_powered
    qg_rho_t = qg_rho.^rho;     % Raise to the power of rho
    
    % Compute the Pi-weighted terms
    pi_g_rho = pi_matrix .* (g_matrix.^(-rho / (1 + rho)));
    
    % Combine results to compute the second component
    component_second = Q' * (pi_g_rho * qg_rho_t');
    
    % Compute the final value for E0(rho)
    Eo_rho = -log2((1 / pi) * component_second);
end